petab/v2/petab1to2.py

```diff
@@ -388,6 +391,41 @@ def update_noise_dist(row):
         df[v2.C.NOISE_DISTRIBUTION] = df.apply(update_noise_dist, axis=1)
         df.drop(columns=[v1.C.OBSERVABLE_TRANSFORMATION], inplace=True)
 
+    def extract_placeholders(row: pd.Series, type_: str) -> str:
+        """Extract placeholders from observable formula."""
+        if type_ == "observable":
+            formula = row[v1.C.OBSERVABLE_FORMULA]
+        elif type_ == "noise":
+            formula = row[v1.C.NOISE_FORMULA]
+        else:
+            raise ValueError(f"Unknown placeholder type: {type_}")
+
+        if pd.isna(formula):
+            return ""
+
+        t = f"{re.escape(type_)}Parameter"
+        o = re.escape(row[v1.C.OBSERVABLE_ID])
+
+        pattern = re.compile(rf"(?:^|\W)({t}\d+_{o})(?=\W|$)")
+
+        expr = sympify_petab(formula)
+        # the current sorting will fail for 10+ placeholders, but who does
+        #  that anyway?
+        return v2.C.PARAMETER_SEPARATOR.join(
+            sorted(
+                str(sym)
+                for sym in expr.free_symbols
+                if sym.is_Symbol and pattern.match(str(sym))
+            )
+        )
```

Comment:
What causes failure for 10+ placeholders?

Reply:
Lexicographical sorting:

```
In[1]: list(sorted(["observableParameter1_foo", "observableParameter10_foo"]))
Out[1]: ['observableParameter10_foo', 'observableParameter1_foo']
```

Well, not the sorting itself will fail, but the placeholder order won't match the override order in the measurement table anymore.